### PR TITLE
[FX] Fix PyTree unpacking carrying forward type annotations

### PR DESCRIPTION
**Pull request by jamesr66a**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #81906

Resolves https://github.com/pytorch/pytorch/issues/81902

### CODE REVIEW

**Review by Chillee — APPROVED** [review]
LGTM.

But I wonder whether we should rethink our pytree integration with FX. Like, our original integration had (essentially) 2 requirements.

1. The FX module was still runnable as-is with the same inputs as before.
2. The FX module is retraceable with `fx.symbolic_trace`.

From my understanding, most of the users of pytrees are now going through `make_fx` - can we drop some of the previous requirements and simplify this code?

File: torch/fx/graph.py [hunk 1]
```diff
@@ -533,17 +545,18 @@ def process_outputs(self, out: Any) -> Any:
...
-        has_orig_self = (function_args[0] == 'self')
+        has_orig_self = (function_args[0].name == 'self')
         if has_orig_self:
-            free_vars.insert(0, 'self')
+            free_vars.insert(0, _ParamDescr('self', '', ''))
         function_definition = super().gen_fn_def(function_args[:], maybe_return_annotation)
         if len(free_vars) > 0:  # pytree has placeholders in it
             function_definition += f"""
-    {', '.join(free_vars)}, = fx_pytree.tree_flatten_spec([{', '.join(function_args)}], self._in_spec)"""
+    {', '.join(var.name for var in free_vars)}, = fx_pytree.tree_flatten_spec\
```
Comment (voznesenskym) [inline torch/fx/graph.py:558]: why trailing comma before `=` ?

**Review by voznesenskym — APPROVED** [review]

File: torch/fx/graph.py (hunk 1 above)
Comment (jamesr66a) [reply torch/fx/graph.py:558]: That forces it to always be a tuple even if there's 1 element

**Comment by voznesenskym** [comment]
@jamesr66a  I saw this got reverted, but we still need this fix. @jeanschmidt how is it breaking internal builds?

**Pull request closed without merging**